### ISSUE TITLE
Support argparse-like functionality

### ISSUE THREAD

**Issue by buddha314** [chapel-lang/chapel#8302]
Based on [Python's Argparse](https://docs.python.org/3/library/argparse.html)

Chapel supports [`main()`](https://chapel-lang.org/docs/master/technotes/main.html) arguments, but does not appear to parse the arguments in any meaningful way.  Something like `argparse` would allow you to do 

```chapel
// oscar.chpl
proc main(args: [] string) {
  var as = parse(args);
  writeln("And the winner is! ", as.winner);
}
```

```bash
./oscar  --winner='Emma Stone'
And the winner is! Emma Stone
```

Right now, it appears you have to loop over `a in args` and parse the string, like splitting on `=`.

**Comment by bradcray** [comment]
Note that if you simply want to parse a `--winner` flag as above, the simplest option would probably be to use a `config const`:

```chapel
// oscar.chpl
config const winner: string;

proc main() {
  writeln("And the winner is! ", winner);
}
```

But I agree that for those programs that want/need more general / complete argv-style argument parsing, something more than dealing with the array of strings manually would be desirable.

**Comment by buddha314** [comment]
Noted.

If it's more "chapel-like" to use `config const` that's fine with me.  I could use examples of how to halt if something isn't defined and stuff like that.  Argparse is pretty mature and has a lot of error catching, etc.

**Comment by buddha314** [comment]
For instance, can I do 

```
// oscar.chpl
config const winner: string;

proc main() {
  if !winner {  // Is 'winner' set?
     writeln("But I waited through all the boring speeches and red carpet stuff!");
  } else {
    writeln("And the winner is! ", winner);
  }
}
```

I suppose I could try this myself, but it seems like the answer might be useful to others, and I'm lazy.

**Comment by bradcray** [comment]
> If it's more "chapel-like" to use config const that's fine with me.

I think for flags that `config const` supports well, that is the more chapeltastic way to do things.  It arguably breaks down if you want to have a command line that lists a bunch of filenames (say) as standalone arguments with no flag defining them.

> Argparse is pretty mature and has a lot of error catching, etc.

I think what you're saying is "I'm used to argparse and think it's pretty nice" but what I hear on my end is "I want you to do a lot of work to recreate all of that capability and complexity for me and have it integrate with `config` declarations naturally."  So I'm inclined to drag my feet until we come up with some compelling cases where `config`s and the current features are problematic.  I'll mention that my preference for handling more arbitrary arguments would be to have a capability in which modules could register argument parsing handlers that would be invoked automatically in a way that reflected the module `use` chain to make it more integrated with the language and program structure, but we haven't found the time to pursue that idea yet either.

I've edited your "nobody defined it" case above to make it work (the only change was to replace `!winner` with `winner == ""`), and you can do that, yes.  Stylistically, I tend to handle such cases closer to their declaration point in order to keep the mainline code cleaner:

```chapel
// oscar.chpl                                                                   
config const winner: string;
assert(winner!="", "But I waited through all the boring speeches and red carpet stuff!");

proc main() {
  writeln("And the winner is! ", winner);
}
```

(or if you don't like the assert's output style, you could use a halt):

```chapel
if winner!="" then halt("But I waited through all the boring speeches and red carpet stuff!");
```

(or a writeln + exit):

```chapel
if winner!="" {
  writeln("But I waited through all the boring speeches and red carpet stuff!");
  exit(1);
}
```

**Comment by ljdursi** [comment]
I generally do, e.g.,

const defaultfilename="SomethingUnlikely";
config const infile=defaultfilename;

if infile != defaultfilename {
...
}

But note that if support for multi line string literals are coming, one needn’t be satisfied with argparse, one could write a docopt package...

**Comment by buddha314** [comment]
They are [HERE](https://github.com/chapel-lang/chapel/issues/7867) and I'm using them already.  Very very useful.

**Comment by ronawho** [comment]
I'm a big fan of argparse, but I agree with Brad that config consts mostly avoid the need for something like argparse.

The only things I find myself missing when I want to write something user-facing are that:
 - It's not easy to attach a help/usage message to a whole program
 - You can't attach a help/info message to individual config const
 - You can't hide non-user facing config consts from the `--help` output (all are printed)
 - You can't have short names (or use unambiguous abbreviations) for variables, you have to type the whole thing out
 
I think most of those are solvable without implementing a full argparse though

**Comment by ben-albrecht** [comment]
> - It's not easy to attach a help/usage message to a whole program
> - You can't attach a help/info message to individual config consts

Somewhat related: #6282

> - You can't hide non-user facing config consts from the --help output (all are printed)

Related: #4973

**Comment by buddha314** [comment]
Even if using config const instead of args, there should be a way to to handle

* Hey, you didn't give me this constant!
* Here are your options for this constant:
   * Option A
   * Option B
* You provided two incompatible flags, idiot.
* The file you said was there?  It wasn't there. I'm using my own default file (as in https://github.com/chapel-lang/chapel/issues/8154)

You can implement each of these by hand, but it's not DRY.  Seems like an easy `use Constants` module or something.

**Comment by ronawho** [comment]
Agreed, those are also useful things that argparse provides. Maybe update the issue message to explicitly list the features of argparse that aren't provided by config consts?

**Comment by cassella** [comment]
I'd add to that list being able to specify a multival option, e.g `gcc -DFOO=bar -D BAZ=SPLAT`, or `sed -f scriptfile1 -f scriptfile2`

**Comment by milthorpe** [comment]
Another nice-to-have feature for `ArgumentParser` to bring it in line with the behaviour of `argparse`: when displaying help, list the arguments in the same order in which they were added (modulo positional arguments, which are always displayed first). 
Currently, the help text lists the arguments in whatever order they fall out of the `_handlers` map. A more friendly user interface would allow the command developer to decide the order in which the arguments are listed.

**Comment by bradcray** [comment]
@milthorpe : Good suggestion!  

Noticing where you posted it:  I think we should close this issue because it's written so broadly that it's not clear what it would take to close it beyond the current ArgParse package (or an arbitrary number of improvements to it).  @arezaii, do you agree?

If we do, we should spin your issue off into its own issue (and maybe Ahmad's "unimplemented features" from the previous comment, if they still stand).

**Comment by arezaii** [comment]
> I think we should close this issue because it's written so broadly that it's not clear what it would take to close it beyond the current ArgParse package (or an arbitrary number of improvements to it)

I agree. I think the original issue is mostly satisfied by the `ArgumentParser` package and any remaining work should be in a specific issue for enhancements. I created https://github.com/chapel-lang/chapel/issues/22900 for this purpose and added @milthorpe's suggestion to the list.

**Comment by bradcray** [comment]
Great, thanks @arezaii!  Closing this on that basis.
